tree/tree/src/TTree.cxx

```diff
@@ -4802,6 +4814,53 @@ Int_t TTree::FlushBaskets() const
    Int_t nerror = 0;
    TObjArray *lb = const_cast<TTree*>(this)->GetListOfBranches();
    Int_t nb = lb->GetEntriesFast();
+
+#ifdef R__USE_IMT
+   if (ROOT::IsImplicitMTEnabled() && fIMTEnabled) {
+      if (fSortedBranches.empty()) { const_cast<TTree*>(this)->InitializeSortedBranches(); }
+
+      BoolRAIIToggle sentry(fIMTFlush);
+      fIMTZipBytes.store(0);
+      fIMTTotBytes.store(0);
+      std::atomic<Int_t> nerrpar(0);
+      std::atomic<Int_t> nbpar(0);
+      std::atomic<Int_t> pos(0);
+      tbb::task_group g;
```

Comment:
Hi Chris,

Indeed.

My logic here is:

- This function is high up enough in the call stack to verify it is not called with any global ROOT mutexes held.
- The locking within the function was made per-`TFile`.
- `TTree` and `TFile` are both thread-unsafe objects (cannot have two threads doing const or non-const calls into the same object), meaning there should not be multiple callers to this function regardless.

So, the remaining danger is if the caller (CMSSW) is holding a mutex of its own that other CMSSW tasks might need.

I will write these concerns up into a comment, provide an example of dangerous behavior, and add them to the function's documentation.

Brian